### PR TITLE
#77 - Update /bid/:bidId/status - Aisha Athman Lali

### PR DESCRIPTION
**Pull request by aishaathmanlali**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

Write a summary of your change. Explain why you have made it.

## Related to

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have carefully reviewed my own code
- [ ] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -71,6 +72,10 @@ const auth = async (req, res, next) => {
 			return next();
 		}
 
+		if (allowedAccess === user.token) {
+			return next();
+		}
+
```
Comment (soonick) [inline server/api.js]: Why is this here? Didn't you already add this in your previous commit?

File: server/api.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ const allowlist = {
 	POST: {
 		"/sign-in": "public",
 		"/tender": "buyer",
+		"/bid": "token",
```
Comment (soonick) [inline server/api.js:13]: Since a request for this endpoint looks like `/bid/123/status` and we want to match only `bid`, we need to split the request path and grab only the first part before matching against our allow list

File: server/api.js [hunk 3]
```diff
@@ -316,50 +322,56 @@ router.post("/bid/:bidId/status", async (req, res) => {
...
+		const bidderId = bid.bidder_id;
+
+		const tenderResult = await client.query(
+			"SELECT buyer_id FROM tender WHERE id = $1;",
+			[tenderId]
+		);
+
+		if (tenderResult.rowCount === 0) {
+			await client.query("ROLLBACK");
+			return res.status(404).send({ code: "TENDER_NOT_FOUND" });
```
Comment (soonick) [inline server/api.js]: This should be a `500` we should always be able to find a tender for a bid

File: server/api.js [hunk 4]
```diff
@@ -316,50 +322,56 @@ router.post("/bid/:bidId/status", async (req, res) => {
...
+		}
+
+		const buyerId = tenderResult.rows[0].buyer_id;
+
+		if (
+			(status === "Awarded" || status === "Rejected") &&
+			user.id !== buyerId
+		) {
+			await client.query("ROLLBACK");
+			return res.status(403).send({ code: "FORBIDDEN" });
```
Comment (soonick) [inline server/api.js:354]: When a bid is awarded, we need to reject all the other bids on the same tender. Why did you remove that code?

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 1 above)
Comment (aishaathmanlali) [reply server/api.js]: I have removed it

File: server/api.js [hunk 5]
```diff
@@ -316,50 +320,78 @@ router.post("/bid/:bidId/status", async (req, res) => {
...
+				[status, bidId]
 			);
 
 			const rejectOtherBidsResult = await client.query(
 				"UPDATE bid SET status = $1 WHERE tender_id = $2 AND bid_id != $3;",
 				[rejectStatus, tenderId, bidId]
 			);
 
-			if (awardBidResult.rowCount > 0 || rejectOtherBidsResult.rowCount > 0) {
+			if (awardBidResult.rowCount > 0 && rejectOtherBidsResult.rowCount >= 0) {
```
Comment (soonick) [inline server/api.js]: `||` is correct. If there is only one bid the rejected bids will be 0

File: server/api.js [hunk 6]
```diff
@@ -316,50 +320,78 @@ router.post("/bid/:bidId/status", async (req, res) => {
...
-				await client.query("COMMIT");
-				return res.status(200).send({ code: "SUCCESS" });
-			} else {
-				await client.query("ROLLBACK");
-				return res.status(500).send({ code: "SERVER_ERROR" });
-			}
+		const updateBidResult = await client.query(
+			"UPDATE bid SET status = $1 WHERE bid_id = $2;",
+			[status, bidId]
+		);
```
Comment (soonick) [inline server/api.js]: Why did you remove the `else`?

**Review by soonick — CHANGES_REQUESTED** [review]

**Review by soonick — APPROVED** [review]
